File: bed-tools-gg/bed_shuffle.R
#!/usr/bin/Rscript

# ------------------------------------------------------------------------------
# 
# Author: Gabriele Girelli
# Email: gigi.ga90@gmail.com
# Version: 1.0.1
# Description: Shuffle a certain percentage of reads in a bed file.
# 
# ------------------------------------------------------------------------------



# DEPENDENCIES =================================================================

suppressMessages(library(argparser))
suppressMessages(library(data.table))
suppressMessages(library(parallel))
suppressMessages(library(readr))

# INPUT ========================================================================

# Create arguent parser
parser = arg_parser('Shuffle bed file reads.', name = 'bed_shuffle.R')

# Define mandatory arguments
parser = add_argument(parser, arg = 'seed', short = '-s',
	help = 'Seed for random number generation.', type = class(0))
parser = add_argument(parser, arg = 'bedfile', short = '-b',
	help = 'path to bedfile.', type = class(''))

# Define elective arguments
parser = add_argument(parser, arg = '--nIter', short = '-n',
	help = 'Number of iterations.',
	default = 100, nargs = 1, type = class(0))
parser = add_argument(parser, arg = '--perc', short = '-p',
	help = 'Percentage of reads to shuffle.',
	default = 10, nargs = 1, type = class(0))
parser = add_argument(parser, arg = '--outDir', short = '-o',
	help = 'Output directory.',
	default = "./shuffled/", nargs = 1, type = class(''))
parser = add_argument(parser, arg = '--keepSeed', short = '-k',
	help = 'Reload previous seed state. Use on subsequent runs.',
	default = TRUE, nargs = 1, type = class(TRUE))
parser = add_argument(parser, arg = '--threads', short = '-t',
	help = 'Number of threads for parallelization.',
	default = 1, nargs = 1, type = class(0))

# Parse arguments
p = parse_args(parser)

# Attach argument values to variables
attach(p['' != names(p)])

fname = unlist(strsplit(bedfile, '/', fixed = TRUE))
fname = fname[length(fname)]
fname = unlist(strsplit(fname, '.', fixed = TRUE))
fname = fname[-length(fname)]
fname = paste(fname, collapse = '.')

# FUNCTIONS ====================================================================

# RUN ==========================================================================

# Log info
cat(paste0(' · Shuffling x', nIter, ' ', perc, '% of ', bedfile, '\n'))

# Set seed
set.seed(seed)

# Load seed if available
fseed = paste0(outDir, '/.Random.seed.RData')
if ( file.exists(fseed) && keepSeed ) {
	cat(paste0(' >> Loading previous seed status...\n'))
	load(fseed)
}

# Read bedfile -----------------------------------------------------------------

bedcols = c('chr', 'start', 'end', 'name', 'score')
bed = read_delim(bedfile, '\t', skip = 1,
	col_names = bedcols, col_types = 'ciici')

# Shuffle ----------------------------------------------------------------------

# Count reads
nreads = sum(bed$score)
toShuffle = round(nreads * perc / 100)
cat(paste0(' >> Found ', nreads, ' reads.\n'))

cat(paste0(' >>> Pre-shuffling...\n'))
preshuffle = factor(unlist(mclapply(seq(nrow(bed)),
	FUN = function(nr) {
		row = bed[nr,]
		return(rep(gsub(' ', '', paste(row[1:4], collapse = '~')), row[5]))
	}
	, mc.cores = threads
)))
cat(paste0(' >>> Identifying locations...\n'))
positions = levels(preshuffle)

l = lapply(seq(nIter),
	function(i) {
		cat(paste0(' >>>> Iteration #', i, '\n'))

		shuffled = preshuffle

		cat(paste0(' >>>> Identifying origins...\n'))
		from = sample(seq(length(preshuffle)), toShuffle)
		cat(paste0(' >>>> Identifying destinations...\n'))
		to = sample(seq(length(positions)), toShuffle, replace = TRUE)

		cat(paste0(' >>>> Moving reads...\n'))
		shuffled[from] = positions[to]

		cat(paste0(' >>>> Counting shuffled reads...\n'))
		counts = table(shuffled)

		cat(paste0(' >>>> Preparing output...\n'))
		out = rbindlist(mclapply(seq(length(counts)),
			FUN = function(j) {
				loc = names(counts)[j]
				loc = unlist(strsplit(loc, '~', fixed = T))
				data = data.frame(
					chr = loc[1],
					start = loc[2],
					end = loc[3],
					name = loc[4],
					score = counts[j]
				)
				return(data)
			}
			, mc.cores = threads
		))
		#cat(paste0(' >> ', sum(out$score), ' reads in the shuffled file.\n'))

		outfname = paste0(outDir, '/', fname, '.iter', i, '.', perc, 'perc.bed')
		write.table(out, outfname, row.names = F, quote = F, sep = '\t')
	}
)

# Save current seed status
cat(paste0(' >> Saving seed status...\n'))
save('.Random.seed', file=paste0(outDir, '/.Random.seed.RData'))

# END --------------------------------------------------------------------------

################################################################################

File: bed-tools-gg/bed_shuffle.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# ------------------------------------------------------------------------------
# 
# Author: Gabriele Girelli
# Email: gigi.ga90@gmail.com
# Version: 1.0.1
# Description: Shuffle a certain percentage of reads in a bed file.
# 
# ------------------------------------------------------------------------------



# DEPENDENCIES =================================================================

import argparse
import numpy as np
import os
import pandas as pd
import pickle
import sys

pd.options.mode.chained_assignment = None  # default='warn'

# INPUT ========================================================================

# Add script description
parser = argparse.ArgumentParser(
	description = 'Shuffle bed file read counts.'
)

# Add params
parser.add_argument('seed', type = int, nargs = 1,
	help = 'Seed for random number generation.')
parser.add_argument('bedfile', type = str, nargs = 1,
	help = 'Path to bedfile.')

# Add flags
parser.add_argument('-n', metavar = 'nIter', type = int, nargs = 1,
	default = [100], help = 'Number of iterations.')
parser.add_argument('-p', metavar = 'perc', type = int, nargs = 1,
	default = [10], help = 'Percentage of reads to shuffle.')
parser.add_argument('-o', metavar = 'outDir', type = str, nargs = 1,
	default = ['./shuffled/'], help = 'Output directory.')
parser.add_argument('-k', metavar = 'keepSeed', type = bool, nargs = 1,
	default = [True],
	help = 'Reload previous seed state. Use on subsequent runs.')

# Parse arguments
args = parser.parse_args()

# Retrieve arguments
seed = args.seed[0]
bedfile = args.bedfile[0]
nIter = args.n[0]
perc = args.p[0]
outDir = args.o[0]
keepSeed = args.k[0]

# Output file name prefix
outName = '.'.join(bedfile.split('/')[-1].split('.')[:-1])

# Make output directory if missing
if not os.path.isdir(outDir):
	os.makedirs(outDir)

# RUN ==========================================================================

# Log info
print(' · Shuffling x'+str(nIter)+' '+str(perc)+'% of '+bedfile)

# Set seed
seed = np.random.RandomState(seed)

# Load seed if available
fname = outDir + '/.seed_state.pickle'
if os.path.isfile(fname):
	f = open(fname, 'r')
	seed.set_state(pickle.load(f))
	f.close()

# Read bedfile -----------------------------------------------------------------

bf = pd.read_csv(bedfile, '\t', skiprows = [0],
	names = ['chr', 'start', 'end', 'name', 'score'])

# Shuffle ----------------------------------------------------------------------

# Count reads
nreads = sum(bf['score'])
toShuffle = nreads * perc / 100

print(' >>> Found ' + str(nreads) + ' reads.')


print(' >>> Pre-shuffling...')
preshuffle = np.repeat(np.arange(len(bf['score'])), bf['score'])

for i in range(nIter):
	print(' >>># Iteration #' + str(i+1))
	print(' >>># Shuffle...')
	pos_from = seed.randint(0, len(preshuffle), toShuffle)
	pos_to = seed.randint(0, len(bf['score']), toShuffle)

	print(' >>># Moving reads...')
	shuffled = np.array(preshuffle)
	shuffled[pos_from] = pos_to

	print(' >>># Counting shuffled reads')
	counts = np.unique(shuffled, return_counts = True)
	shuffled = bf.copy()
	shuffled['score'] = np.zeros(shuffled['score'].shape)
	shuffled['score'][counts[0]] = counts[1]

	# Output
	shuffled.to_csv(outDir+outName+'.iter'+str(i+1)+'.'+str(perc)+'perc.bed',
		sep = '\t', header = False, index = False)

# Saving seed state
seed_state = seed.get_state()
f = open(outDir + '/.seed_state.pickle', 'w+')
pickle.dump(seed_state, f)
f.close()

# END --------------------------------------------------------------------------

################################################################################
